core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java

```diff
@@ -260,31 +297,41 @@ public synchronized void close()
         }
     }
 
-    private synchronized void scheduleRequestIfNecessary()
+    @VisibleForTesting
+    synchronized int scheduleRequestIfNecessary()
     {
         if ((buffer.isFinished() || buffer.isFailed()) && completedClients.size() == allClients.size()) {
-            return;
+            return 0;
         }
 
         long neededBytes = buffer.getRemainingCapacityInBytes();
         if (neededBytes <= 0) {
-            return;
+            return 0;
         }
 
-        int clientCount = (int) ((1.0 * neededBytes / averageBytesPerRequest) * concurrentRequestMultiplier);
-        clientCount = Math.max(clientCount, 1);
-
-        int pendingClients = allClients.size() - queuedClients.size() - completedClients.size();
-        clientCount -= pendingClients;
+        long reservedBytesForScheduledClients = allClients.values().stream()
+                .filter(client -> !queuedClients.contains(client) && !completedClients.contains(client))
+                .mapToLong(HttpPageBufferClient::getAverageRequestSizeInBytes)
+                .sum();
+        long bytesToBeRequested = 0;
+        int clientCount = 0;
 
+        for (HttpPageBufferClient client : queuedClients) {
+            if (bytesToBeRequested > neededBytes * concurrentRequestMultiplier - reservedBytesForScheduledClients) {
```

Comment:
why not `>=`? What is the point of making a request when `reservedBytesForScheduledClients` is already ok. I think it worked like that previously `clientCount -= pendingClients;`